malsim/sims/mal_simulator.py

```diff
@@ -103,42 +103,63 @@ def create_blank_observation(self, default_obs_state=-1):
         # For now, an `object` is an attack step
         num_steps = len(self.attack_graph.nodes)
 
+        # Since nodes are stored as sets and we need deterministic
+        # observations in the simulator, we need to convert them
+        # to a list and sort them
+        ordered_nodes: list[AttackGraphNode] = (
+            list(self.attack_graph.nodes.values())
+        )
+        ordered_nodes.sort(key=lambda n: n.id)
+
         observation = {
             # If no observability set for node, assume observable.
             "is_observable": [step.extras.get('observable', 1)
-                           for step in self.attack_graph.nodes],
+                           for step in ordered_nodes],
             # Same goes for actionable.
             "is_actionable": [step.extras.get('actionable', 1)
-                           for step in self.attack_graph.nodes],
+                           for step in ordered_nodes],
             "observed_state": num_steps * [default_obs_state],
             "remaining_ttc": num_steps * [0],
-            "asset_type": [self._asset_type_to_index[step.asset.type]
-                           for step in self.attack_graph.nodes],
-            "asset_id": [step.asset.id
-                         for step in self.attack_graph.nodes],
-            "step_name": [self._step_name_to_index[
-                          str(step.asset.type + ":" + step.name)]
-                          for step in self.attack_graph.nodes],
+            "asset_type": [self._asset_type_to_index[step.lg_attack_step.asset.name]
+                           for step in ordered_nodes],
+            "asset_id": [step.model_asset.id
+                         for step in ordered_nodes],
+            "step_name": [
+                self._step_name_to_index.get(
+                    str(step.lg_attack_step.asset.name + ":" + step.name)
+                ) for step in ordered_nodes],
```

Comment:
I'm not sure which one is faster sorting the dictionary and then having these lists created like they are right now or just iterating through the keys and appending to the lists one element at a time.